drivers: gpio: xlnx: Add interrupt support

robhancocksed opened the pull request:
Added support for interrupt-driven GPIO for this driver if an interrupt line is defined. Note that interrupt support is optional in the GPIO core so it may or may not be present.

The core provides a single state change notification whenever any line changes state on a given bank, so we need to check the states of the lines ourselves to determine what has actually changed.

Also, the driver treats the first and second GPIO banks as mostly separate devices, but a single interrupt is shared by both banks, so the ISR ends up being registered by the first bank which also triggers handling on the second bank if required.

robhancocksed committed:
drivers: gpio: xlnx: Add interrupt support

Added support for interrupt-driven GPIO for this driver if an interrupt
line is defined. Note that interrupt support is optional in the GPIO
core so it may or may not be present.

The core provides a single state change notification whenever any line
changes state on a given bank, so we need to check the states of the
lines ourselves to determine what has actually changed.

Also, the driver treats the first and second GPIO banks as mostly
separate devices, but a single interrupt is shared by both banks, so the
ISR ends up being registered by the first bank which also triggers
handling on the second bank if required.

Signed-off-by: Robert Hancock <robert.hancock@calian.com>

robhancocksed commented:
Pushed a change to the DT schema to fix a compliance error.

henrikbrixandersen reviewed (changes requested):
Please make all these additions compile-time #ifdef'ed based on whether any instance has an interrupt line configured. You can use https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/can/transceiver/can_transceiver_gpio.c#L16 as inspiration.

robhancocksed committed:
drivers: gpio: xlnx: Fix property reading for second channel

The properties being read for the all-inputs and all-outputs flags in
the driver for the second GPIO channel did not match the ones listed in
the DT schema or the existing device tree files. Fix this.

Signed-off-by: Robert Hancock <robert.hancock@calian.com>

robhancocksed commented:
> Please make all these additions compile-time #ifdef'ed based on whether any instance has an interrupt line configured. You can use https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/can/transceiver/can_transceiver_gpio.c#L16 as inspiration.

I've added this in to make the interrupt support conditional. I also noticed another issue where some of the DT properties being read for the second channel were incorrect, I've added in a fix for that in a separate commit.

robhancocksed commented:
Rebased onto the updated main branch.

robhancocksed commented:
@henrikbrixandersen I think your requested changes have been addressed now.

henrikbrixandersen commented on drivers/gpio/gpio_xlnx_axi.c:
```diff
@@ -193,38 +210,151 @@ static int gpio_xlnx_axi_port_toggle_bits(const struct device *dev,
+static int gpio_xlnx_axi_pin_interrupt_configure(const struct device *dev, gpio_pin_t pin,
+						 enum gpio_int_mode mode, enum gpio_int_trig trig)
 {
-	ARG_UNUSED(dev);
-	ARG_UNUSED(pin);
-	ARG_UNUSED(mode);
-	ARG_UNUSED(trig);
```
You need to preserve these in the `#else` case.

robhancocksed replied:
Re-added.

henrikbrixandersen commented on drivers/gpio/gpio_xlnx_axi.c:
```diff
@@ -193,38 +210,151 @@ static int gpio_xlnx_axi_port_toggle_bits(const struct device *dev,
+static int gpio_xlnx_axi_manage_callback(const struct device *dev, struct gpio_callback *callback,
 					 bool set)
 {
-	ARG_UNUSED(dev);
-	ARG_UNUSED(cb);
-	ARG_UNUSED(set);
```
Same here.

robhancocksed replied:
Re-added.

henrikbrixandersen commented on drivers/gpio/gpio_xlnx_axi.c:
```diff
@@ -233,6 +363,25 @@ static int gpio_xlnx_axi_init(const struct device *dev)
+#if ANY_INST_HAS_INTERRUPTS
+	const struct gpio_xlnx_axi_config *config = dev->config;
+
+	if (config->irq_config_func) {
```
```suggestion
	if (config->irq_config_func != NULL) {
```

robhancocksed replied:
changed

henrikbrixandersen commented on drivers/gpio/gpio_xlnx_axi.c:
```diff
@@ -248,75 +397,78 @@ static const struct gpio_driver_api gpio_xlnx_axi_driver_api = {
 	.get_pending_int = gpio_xlnx_axi_get_pending_int,
 };
 
-#define GPIO_XLNX_AXI_GPIO2_HAS_COMPAT_STATUS_OKAY(n)			\
```
It's very difficult to tell what is being changed in the code below since your commit contains both functional and whitespace changes. Please separate the two or leave out the whitespace changes all together.

robhancocksed replied:
I split the whitespace changes out into a separate commit.

henrikbrixandersen commented on dts/bindings/gpio/xlnx,xps-gpio-1.00.a.yaml:
```diff
@@ -13,6 +13,9 @@ properties:
   reg:
     required: true
 
+  interrupts:
+    description: Interrupt for GPIO core, if enabled
+
```
This is already described in `base.yaml`. If it is to be included here, the description needs to be much more detailed and specific to this IP core. Otherwise, please remove it.

robhancocksed replied:
Removed

henrikbrixandersen reviewed (changes requested):
Apologies for the long delay in reviewing. Somehow this had fallen off my list.

robhancocksed committed:
drivers: gpio: xlnx: run clang-format

Clean up whitespace etc. using clang-format after previous changes. No
functional changes intended.

Signed-off-by: Robert Hancock <robert.hancock@calian.com>

robhancocksed commented on drivers/gpio/gpio_xlnx_axi.c:
```diff
@@ -32,12 +33,20 @@
 #define MAX_GPIOS 32
 
+#define INST_HAS_INTERRUPTS_OR(inst) DT_INST_NODE_HAS_PROP(inst, interrupts) ||
+#define ANY_INST_HAS_INTERRUPTS DT_INST_FOREACH_STATUS_OKAY(INST_HAS_INTERRUPTS_OR) 0
```
Added.

henrikbrixandersen reviewed (approved):
Thanks!
